composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+        elif self.split == "validation":
+            self.num_shards = 8
+            self.approx_samples_per_shard = 45576
+        else:
+            raise ValueError(f"Unknown {split=}, expected 'train' or 'validation'.")
+
+        # Set dataset size
+        self.world_size = dist.get_world_size()
+        self.rank = dist.get_global_rank()
+        self.max_samples_per_device = self.max_samples // self.world_size
+        new_max_samples = self.max_samples_per_device * self.world_size
+        if new_max_samples != self.max_samples:
+            log.warning(
+                f"Max samples will be truncated from {max_samples}->{new_max_samples} to maintain divisibility across {self.world_size} devices."
+            )
+            self.max_samples = new_max_samples
+
+        # Try and detect if max_samples is too large
+        # TODO: For safety, repeat original dataset if max_samples is larger than original size
+        original_approx_samples = self.num_shards * self.approx_samples_per_shard
+        if self.max_samples > original_approx_samples and self.group_method == "truncate":
+            raise ValueError(
+                f"Max samples was set to {self.max_samples} with group_method 'truncate' but split '{split}' has only {original_approx_samples}.)"
+            )
+        if self.group_method == "concat":
+            log.warning(
+                f"When using group_method 'concat', sequential token samples are concatenated and chunked into fixed-length samples of size {self.max_seq_len=}. "
+                f"In general we cannot detect ahead-of-time if your setting of {self.max_samples=} is safe, "
+                f"so please be cautious to not request more tokens than the size of the original dataset.")
+
+        # Build tokenizer
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_name)
+        if self.tokenizer_name in ["gpt2"]:
+            # The GPT2 tokenizer has no padding token which causes bugs
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+
+        # Load and shard dataset
+        text_dataset = datasets.load_dataset(path="c4", name="en", split=split, streaming=True)
+        text_dataset = self._shard_dataset(text_dataset)
+        if not isinstance(text_dataset, datasets.IterableDataset):
+            raise ValueError("Unable to build sharded Huggingface C4 Dataset.")
+
+        # Shuffle dataset
+        # Per-device shard list is shuffled at the start of each epoch
+        # Samples are read into and randomly sampled from per-device shuffle buffer
+        if self.shuffle:
+            text_dataset = text_dataset.shuffle(buffer_size=self.shuffle_buffer_size, seed=self.seed)
+
+        # Map text samples to token samples
+        # NOTE: Mapping is executed in batched mode for better CPU utilization,
+        # but the returned dataset is still an iterable over tokenized samples
+        text_sample_batch_size = 1000
```

Comment:
Can we make this an hparam of the dataset?

Reply:
I'm a little wary of exposing this as it's kind of a CPU utilization thing and seems like clutter, I think it's also the HF default (so maybe I can even remove it from my code). What do you think?

Reply:
Closing for now, I think it will be simpler to keep this hparam internal as it's a perf thing.